### PR TITLE
issue #2535, added authentication header for GO

### PR DESCRIPTION
**Pull request by guohuang**
link #2535, i am not sure what `GetApiKeyWithPrefix` is, so i did not implement it, let me know if you still want this, and please explain how

@wing328 please review

### CODE REVIEW

**Comment by wing328** [comment]
@guohuang  `GetApiKeyWithPrefix` is for adding a prefix to the authentication API key, e.g.
```
AUTHORIZATION: TOKEN API_KEY_HERE
```
As you can see sometimes the authentication needs a prefix (TOKEN in this case)

so in configuration struct, ApiKey should be a map of string and we need to add ApiKeyPrefix (map of string) as another property in the configuration struct.

File: modules/swagger-codegen/src/main/resources/go/api.mustache [hunk 1]
```diff
@@ -43,6 +43,28 @@ func (a {{classname}}) {{nickname}} ({{#allParams}}{{paramName}} {{{dataType}}}{
 
     _sling := sling.New().{{httpMethod}}(a.Configuration.BasePath)
 
+    {{#authMethods}}// authentication ({{name}}) required
+    {{#isApiKey}}{{#isKeyInHeader}}
+    // set key with prefix in header
+    _sling.Set("{{keyParamName}}", a.Configuration.GetApiKeyWithPrefix("{{keyParamName}}")
+    {{/isKeyInHeader}}{{#isKeyInQuery}}
+    // set key with prefix in querystring
+    _sling.Set("{{keyParamName}}",  a.Configuration.GetApiKeyWithPrefix("{{keyParamName}}")
```
Comment (guohuang) [inline modules/swagger-codegen/src/main/resources/go/api.mustache]: @wing328 i haven't figure out how to set queryparam using sling, it seems complicated, i have to generate a struct and pass the parameter there using `QueryStruct`

Comment (wing328) [inline modules/swagger-codegen/src/main/resources/go/api.mustache]: Yes, you need to use the QueryParams struct: https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/go/api.mustache#L58-L63

File: modules/swagger-codegen/src/main/resources/go/api.mustache [hunk 2]
```diff
@@ -43,6 +43,32 @@ func (a {{classname}}) {{nickname}} ({{#allParams}}{{paramName}} {{{dataType}}}{
 
     _sling := sling.New().{{httpMethod}}(a.Configuration.BasePath)
 
+    {{#authMethods}}// authentication ({{name}}) required
+    {{#isApiKey}}{{#isKeyInHeader}}
+    // set key with prefix in header
+    _sling.Set("{{keyParamName}}", a.Configuration.GetApiKeyWithPrefix("{{keyParamName}}")
+    {{/isKeyInHeader}}{{#isKeyInQuery}}
+    // set key with prefix in querystring
+    {{#hasKeyParamName}}    type KeyQueryParams struct {
```
Comment (guohuang) [inline modules/swagger-codegen/src/main/resources/go/api.mustache]: @wing328 code has been updated, but there is no sample that uses query param, so i am not 100% sure if the output will be correct.

Comment (wing328) [inline modules/swagger-codegen/src/main/resources/go/api.mustache]: @guohuang thanks. I'll perform a test manually to confirm

File: modules/swagger-codegen/src/main/resources/go/api.mustache [hunk 3]
```diff
@@ -43,6 +43,32 @@ func (a {{classname}}) {{nickname}} ({{#allParams}}{{paramName}} {{{dataType}}}{
 
     _sling := sling.New().{{httpMethod}}(a.Configuration.BasePath)
 
+    {{#authMethods}}// authentication ({{name}}) required
+    {{#isApiKey}}{{#isKeyInHeader}}
+    // set key with prefix in header
+    _sling.Set("{{keyParamName}}", a.Configuration.GetApiKeyWithPrefix("{{keyParamName}}")
```
Comment (wing328) [inline modules/swagger-codegen/src/main/resources/go/api.mustache]: it's missing a ")" at the end

**Comment by guohuang** [comment]
@wing328 added the missing ")" thanks for checking

**Comment by wing328** [comment]
@guohuang thanks. I'll take another look and merge into master if the test result is good.

**Comment by wing328** [comment]
@guohuang if the query parameter/apiKey starts with lower case letter, it's pending the PR: https://github.com/swagger-api/swagger-codegen/pull/2568

File: modules/swagger-codegen/src/main/resources/go/configuration.mustache [hunk 4]
```diff
@@ -1,32 +1,53 @@
...
+}
+
+func (c *Configuration) GetApiKeyWithPrefix(apiKeyIdentifier string) string {
+    var returnValue = c.ApiKey[apiKeyIdentifier]
+    var apiKeyPrefix = c.ApiKeyPrefix[apiKeyIdentifier]
+    if apiKeyPrefix != ""{
+        returnValue = apiKeyPrefix + " " + returnValue
+    }
+        
+    return returnValue
```
Comment (wing328) [inline modules/swagger-codegen/src/main/resources/go/configuration.mustache]: @guohuang Suggestion: I think we can skip the local variable `returnValue` and `apiKeyPrefix` to reduce the overhead in declaring additional local variables.

Comment (guohuang) [inline modules/swagger-codegen/src/main/resources/go/configuration.mustache]: ok, i will submit a PR later

**Pull request merged**